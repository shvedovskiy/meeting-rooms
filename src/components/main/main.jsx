// @flow
import * as React from 'react';
import { connect } from 'react-redux';

import type { State, Dispatch } from '../../store/types';
import type { Counter } from '../../store/counter/types';
import * as actions from '../../store/counter/actions';
import { getCounter } from '../../store/root-selectors';

import classes from './main.module.css';
import logo from './logo.svg';

type Props = {
  counter: Counter,
  incrementCounter: () => void,
  decrementCounter: () => void,
  incrementCounterAsync: number => void,
};

export const Main = ({
  counter,
  incrementCounter,
  decrementCounter,
  incrementCounterAsync,
}: Props) => (
  <div className={classes.Main}>
    <header className={classes.Main_header}>
      <img src={logo} className={classes.Main_logo} alt="logo" />
      <p>
        Edit <code>src/App.js</code> and save to reload.
      </p>
      <a
        href="https://reactjs.org"
        className={classes.Main_link}
        target="_blank"
        rel="noopener noreferrer"
      >
        Learn React
      </a>
    </header>
    <main className={classes.Main_content}>
      <div className={classes.Main_counter}>
        <p>Counter: {counter}</p>
      </div>
      <div className={classes.Main_buttons}>
        <button onClick={() => incrementCounter()}>+1</button>
        <button onClick={() => decrementCounter()}>-1</button>
        <button onClick={() => incrementCounterAsync(Math.random() * 5)}>
          +1 Async
        </button>
      </div>
    </main>
  </div>
);

const mapStateToProps = (state: State) => ({
  counter: getCounter(state),
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  incrementCounter() {
    return dispatch(actions.increment());
  },
  decrementCounter() {
    return dispatch(actions.decrement());
  },
  incrementCounterAsync(delay: number) {
    return dispatch(actions.incrementAsync(delay));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Main);
