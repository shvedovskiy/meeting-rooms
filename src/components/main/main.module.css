.Main {
  text-align: center;
  height: 100vh;
}

.Main_logo {
  animation: Main-logo-spin infinite 20s linear;
  height: 40vmin;
}

.Main_header {
  background-color: #282c34;
  height: 85%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.Main_link {
  color: #61dafb;
}

.Main_content {
  height: 15%;
  display: flex;
  align-items: center;
}

.Main_counter {
  flex-grow: 1;
  margin-right: 20vmin;
}

.Main_buttons {
  flex-grow: 1;
}

.Main_buttons > button:not(:last-child) {
  margin-right: 7px;
}

@keyframes Main-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
