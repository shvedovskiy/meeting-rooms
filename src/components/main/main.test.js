import React from 'react';

import { Main } from './main';

const setup = (counter = 0) => {
  const actions = {
    incrementCounter: jest.fn(),
    decrementCounter: jest.fn(),
    incrementCounterAsync: jest.fn(),
  };
  const wrapper = global.shallow(<Main counter={counter} {...actions} />);

  return {
    wrapper,
    actions,
    buttons: wrapper.find('button'),
    p: wrapper.find('main p'),
  };
};

describe('<Main />', () => {
  it('renders without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should display count', () => {
    const { p } = setup();
    expect(p.text()).toMatch(/^Counter: 0/);
  });

  it('first button should call incrementCounter', () => {
    const { buttons, actions } = setup();
    buttons.at(0).simulate('click');
    expect(actions.incrementCounter).toBeCalled();
  });

  it('second button should call decrementCounter', () => {
    const { buttons, actions } = setup();
    buttons.at(1).simulate('click');
    expect(actions.decrementCounter).toBeCalled();
  });

  it('third button should call incrementCounterAsync', () => {
    const { buttons, actions } = setup();
    buttons.at(2).simulate('click');
    expect(actions.incrementCounterAsync).toBeCalled();
  });
});
