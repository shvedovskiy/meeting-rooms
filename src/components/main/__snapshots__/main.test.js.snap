// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Main
    counter={0}
    decrementCounter={[MockFunction]}
    incrementCounter={[MockFunction]}
    incrementCounterAsync={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="Main_header"
        >
          <img
            alt="logo"
            className="Main_logo"
            src="logo.svg"
          />
          <p>
            Edit 
            <code>
              src/App.js
            </code>
             and save to reload.
          </p>
          <a
            className="Main_link"
            href="https://reactjs.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            Learn React
          </a>
        </header>,
        <main
          className="Main_content"
        >
          <div
            className="Main_counter"
          >
            <p>
              Counter: 
              0
            </p>
          </div>
          <div
            className="Main_buttons"
          >
            <button
              onClick={[Function]}
            >
              +1
            </button>
            <button
              onClick={[Function]}
            >
              -1
            </button>
            <button
              onClick={[Function]}
            >
              +1 Async
            </button>
          </div>
        </main>,
      ],
      "className": "Main",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="logo"
              className="Main_logo"
              src="logo.svg"
            />,
            <p>
              Edit 
              <code>
                src/App.js
              </code>
               and save to reload.
            </p>,
            <a
              className="Main_link"
              href="https://reactjs.org"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn React
            </a>,
          ],
          "className": "Main_header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "logo",
              "className": "Main_logo",
              "src": "logo.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Edit ",
                <code>
                  src/App.js
                </code>,
                " and save to reload.",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Edit ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "src/App.js",
                },
                "ref": null,
                "rendered": "src/App.js",
                "type": "code",
              },
              " and save to reload.",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Learn React",
              "className": "Main_link",
              "href": "https://reactjs.org",
              "rel": "noopener noreferrer",
              "target": "_blank",
            },
            "ref": null,
            "rendered": "Learn React",
            "type": "a",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="Main_counter"
            >
              <p>
                Counter: 
                0
              </p>
            </div>,
            <div
              className="Main_buttons"
            >
              <button
                onClick={[Function]}
              >
                +1
              </button>
              <button
                onClick={[Function]}
              >
                -1
              </button>
              <button
                onClick={[Function]}
              >
                +1 Async
              </button>
            </div>,
          ],
          "className": "Main_content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p>
                Counter: 
                0
              </p>,
              "className": "Main_counter",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Counter: ",
                  0,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Counter: ",
                0,
              ],
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  onClick={[Function]}
                >
                  +1
                </button>,
                <button
                  onClick={[Function]}
                >
                  -1
                </button>,
                <button
                  onClick={[Function]}
                >
                  +1 Async
                </button>,
              ],
              "className": "Main_buttons",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "+1",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "+1",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "-1",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "-1",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "+1 Async",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "+1 Async",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="Main_header"
          >
            <img
              alt="logo"
              className="Main_logo"
              src="logo.svg"
            />
            <p>
              Edit 
              <code>
                src/App.js
              </code>
               and save to reload.
            </p>
            <a
              className="Main_link"
              href="https://reactjs.org"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn React
            </a>
          </header>,
          <main
            className="Main_content"
          >
            <div
              className="Main_counter"
            >
              <p>
                Counter: 
                0
              </p>
            </div>
            <div
              className="Main_buttons"
            >
              <button
                onClick={[Function]}
              >
                +1
              </button>
              <button
                onClick={[Function]}
              >
                -1
              </button>
              <button
                onClick={[Function]}
              >
                +1 Async
              </button>
            </div>
          </main>,
        ],
        "className": "Main",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="logo"
                className="Main_logo"
                src="logo.svg"
              />,
              <p>
                Edit 
                <code>
                  src/App.js
                </code>
                 and save to reload.
              </p>,
              <a
                className="Main_link"
                href="https://reactjs.org"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn React
              </a>,
            ],
            "className": "Main_header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "logo",
                "className": "Main_logo",
                "src": "logo.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Edit ",
                  <code>
                    src/App.js
                  </code>,
                  " and save to reload.",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Edit ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "src/App.js",
                  },
                  "ref": null,
                  "rendered": "src/App.js",
                  "type": "code",
                },
                " and save to reload.",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Learn React",
                "className": "Main_link",
                "href": "https://reactjs.org",
                "rel": "noopener noreferrer",
                "target": "_blank",
              },
              "ref": null,
              "rendered": "Learn React",
              "type": "a",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="Main_counter"
              >
                <p>
                  Counter: 
                  0
                </p>
              </div>,
              <div
                className="Main_buttons"
              >
                <button
                  onClick={[Function]}
                >
                  +1
                </button>
                <button
                  onClick={[Function]}
                >
                  -1
                </button>
                <button
                  onClick={[Function]}
                >
                  +1 Async
                </button>
              </div>,
            ],
            "className": "Main_content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p>
                  Counter: 
                  0
                </p>,
                "className": "Main_counter",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Counter: ",
                    0,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "Counter: ",
                  0,
                ],
                "type": "p",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    onClick={[Function]}
                  >
                    +1
                  </button>,
                  <button
                    onClick={[Function]}
                  >
                    -1
                  </button>,
                  <button
                    onClick={[Function]}
                  >
                    +1 Async
                  </button>,
                ],
                "className": "Main_buttons",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "+1",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "+1",
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "-1",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "-1",
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "+1 Async",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "+1 Async",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
